rfcs/0190-queue-change-task-run-priority.md

```diff
@@ -0,0 +1,55 @@
+# RFC 0190 - Queue change task run priority
+* Comments: [#0190](https://github.com/taskcluster/taskcluster-rfcs/pull/190)
+* Proposed by: @lotas
+
+## Summary
+
+Since we refactored queue internals it became possible to see what tasks are pending for a particular
+worker pool / task queue.
+
+This RFC proposes new API method that would allow to change the priority of the existing task run.
+
+## Motivation
+
+There might be several use-cases when some worker pool has a lot of pending tasks and there is a need
+to move some more important task to the top of the waiting list, or also move down less important ones.
+
+Instead of cancelling existing tasks we can instead choose to change task run priorities.
+
+## Details
+
+I propose we implement this on a task run level and not the task itself:
+
+* tasks are supposed to be immutable
+* changing the priority makes more sense in the context of a specific run, not the whole task
+
+Queue service will expose new method `queue.changeTaskRunPriority(taskId, runId, newPriority)`.
+
+This will try to do the following:
+
+* update `task.runs[runId]` in the database to reflect priority change (possibly extra field to keep the original)
+* update `priority` field in the `queue_pending_tasks` table
+
+If any of this fails (for example task is being changed by some other process), this method will also fail.
+
+This will also require introduction of the new scopes (anyOf):
+
+* `queue:change-task-run-priority-in-queue:<taskQueueId>`
+* `queue:change-task-run-priority:<taskId>`
+
+
+Current order for picking up tasks is based on the task priority and insertion time (FIFO).
+This RFC proposes to change the priority only, and leave the insertion time as is.
+
+
+### Other considerations
+
+Chain of Trust validation might be broken if we change too much of the task definition or run.
```

Comment:
I _believe_ that CoT is unaffected by anything in the `run` other than artifacts. As far as I can tell, we only pull the task definition and the latest artifacts. (This makes sense, because it's job is compare the expected task definition with what was retrieved at scriptworker runtime - there's no way for us to generate an "expected" run.)

If this is wrong for some reason it's probably OK to have CoT ignore priority when verifying things (but may require an RRA update).

Reply:
thanks Ben, I think we would need to update the task definition in light of recent comments. 
As I see it is more important to control whole task group's priority rather than individual tasks (or both equally likely to happen) 

To implement it on the group level and take into account not yet scheduled tasks, we can:
* simply update task definition to change the priority part of it (CoT might need to account for this situations)
* store the updated priority in a separate column, which will be used for scheduling tasks (while keeping original definition unchanged)